### PR TITLE
feat: atlas pull and push scripts | FC-55

### PR DESCRIPTION
**Pull request by Amr-Nash**
### Description
This pull request provides the CI scripts and developer tooling for the [Atlas Translations Management Design
](https://github.com/openedx/openedx-app-ios/blob/develop/docs/0002-atlas-translations-management.rst) proposal related to [OEP-58](https://docs.openedx.org/en/latest/developers/concepts/oep58.html).

#### Testing the `pull_translations`: 

```
make ATLAS_OPTIONS='--repository=Zeit-Labs/openedx-translations --revision=fc_55_sample' pull_translations
```

##### When the above command is executed:

1) The None-English translations will be pulled from openedx-translations repo into `I18N/` directory.
2) Then each language file will be split and distributed to the modules one by one
3) In the end, the pulled translations will be removed and only the split translations can be found in the modules' directories

##### If the script is working as intended:
1) The above command should not produce errors when executed.
2) The pulled translations should be correctly distributed among the modules in the expected paths.
3) The I18N directory should not exist after the script has been executed.
4) The app should have all its translations in place and can now be deployed.

#### Testing the `extract_translations`:

```
make extract_translations
```

##### When the above command is executed:

1) The English translations from all `en.lproj/Localizable.strings` files in all modules will be extracted
2) The keys of each translation entry will be updated to contain the name of the module followed by a dot e.g. `"Module_name.OLD_KEY_of_the_entry"`.
3) Then all of the updated entries will be put in one file: `I18N/en.lproj/Localizable.strings` 

##### If this script is working as intended:
1) The above command should not produce errors when executed.
2) Every entry from the English translation of each module should exist in `I18N/en.lproj/Localizable.strings` with the 3) corresponding comment, if present, placed before it.
3) The keys in I18N should contain the name of the module from which the entry originated, as specified.
4) The values and comments should be identical to the original ones.

### Status

We'll merge to this pull request into this branch multiple pull requests. We'll keep this open until we're ready for review.

This pull request is useful to watch progress.

#### 1. When are source strings extracted? what do the files look like?

  - An automated action will be developed to extract the English sources if they are changed in the IOS repo, then the action will push the sources to the translations repo.
  - The sources will be in one file: `I18N/en.lproj/Localizable.strings` and it will look like this:
```
"Module_one.FIRST.ENTRY.KEY.IN.MODULE.ONE" = "English Translation for the first entry in module one.";
"Module_one.SECOND.ENTRY.KEY.IN.MODULE.ONE" = "English Translation for the second entry in module one.";
.
.
"Module_two.FIRST.ENTRY.KEY.IN.MODULE.TWO" = "Translation for the first entry in module two";
.
.
```

#### 2. Does anything happen to those files before they are added to openedx-translations
The only extracted file is the English sources from the modules in the IOS repo as it is the only translation that lives in that repo. 
TBC

#### 3. When translation files are pulled via atlas what do the files look like?
When the translation files are pulled via atlas the files' structure is going to look like the structure below:
```
I18N/
----ar.lproj/
--------Localizable.strings
----uk.lproj/
--------Localizable.strings
----es.lproj/
--------Localizable.strings
```
As an example the es.lproj/Localiztion.strings would look like:
`I18N/es.lproj/Localizable.strings` 
```
"Module_one.FIRST.ENTRY.KEY.IN.MODULE.ONE" = "Traducción al inglés de la primera entrada del módulo uno.";
"Module_one.SECOND.ENTRY.KEY.IN.MODULE.ONE" = "Traducción al inglés de la segunda entrada del módulo uno.";
.
.
"Module_two.FIRST.ENTRY.KEY.IN.MODULE.TWO" = "Traducción de la primera entrada del módulo dos.";
.
.
```
#### 4) What is done to those files after `atlas pull` to make them work in the app build process?

After `atlas pull` is completed, all language translation files are retrieved, similar to the process described in the previous question. Subsequently, a Python script is executed to process each file. This script iterates through each translation file, splits it, removes the module name from the keys, and organizes each entry into its corresponding module.

For instance, the `I18N/es.lproj/Localizable.strings` file is split into two separate files:

1. `Module_one/Module_one/es.lproj/Localizable.strings`
2. `Module_two/Module_two/es.lproj/Localizable.strings`

Inside `Module_one/Module_one/es.lproj/Localizable.strings`, the content would resemble:
```
"FIRST.ENTRY.KEY.IN.MODULE.ONE" = "English translation of the first entry in module one.";
"SECOND.ENTRY.KEY.IN.MODULE.ONE" = "English translation of the second entry in module one.";
.
.
.
```

Similarly, inside `Module_two/Module_two/es.lproj/Localizable.strings`, it would appear as:
```
"FIRST.ENTRY.KEY.IN.MODULE.TWO" = "English translation of the first entry in module two.";
.
.
.
```

Once the Python script completes its task and the files are split, the make script proceeds to remove the entire `I18N` directory and its contents.

#### 5) Why is this needed?
This new process is being introduced to have the best combination of Developer Experience and Translator Experience.

The best experience for Translators requires combining source strings into as few transifex resources as possible. The best experience for Engineers requires splitting translation source files to fit within the modular architecture.

#### 6) What should live in [openedx translations](https://github.com/openedx/openedx-translations)?
Translations that were done by the open-edx translators are going to be stored in  [openedx translations](https://github.com/openedx/openedx-translations).

#### 7) What should live in atlas?
Atlas is the tool that we are using to pull the translations from [openedx translations](https://github.com/openedx/openedx-translations); therefor, no translations are stored in atlas.

#### 8) What should live in the [openedx-app-ios](https://github.com/openedx/openedx-app-ios) repo?
In the IOS app repo only the English sources are going to be stored.

### TODO

 - [x] Fix the `python3-localizable` pip package for python3 compatibility
 - [x] Install `openedx-atlas` in the virtual env
 - [x] Refactor the code into a single Python script to improve code reuse
 - [ ] Add CI scripts for `make extract_translations`
 - [ ] Test with the iOS team (Volo and others)
 - [x] Review by Brian Smith

### Related issues
 - Closes https://github.com/openedx/openedx-app-ios/issues/106

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @Amr-Nash! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.




:warning: We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](https://openedx.org/cla)     or indicated your institutional affiliation. Please see the [CONTRIBUTING](https://github.com/openedx/.github/blob/master/CONTRIBUTING.md)     file for more information.     If you've signed an agreement in the past, you may need to re-sign.  See     [The New Home of the Open edX Codebase](https://open.edx.org/blog/the-new-home-of-the-open-edx-codebase/)     for details.

    
Once you've signed the CLA, please allow 1 business day for it to be processed.  After this time, you can re-run the     CLA check by adding a comment here that you have signed it.  If the problem persists, you can tag the `@openedx/cla-problems` team in a     comment on your PR for further assistance.

**Comment by OmarIthawi** [comment]
Note: CLA has been submitted but needs a day or two to reflect on GitHub.

**Review by OmarIthawi — COMMENTED** [review]
Thanks @Amr-Nash, this pull request needs the following:

 - squash the commits
 - use [conventional commits](https://open-edx-proposals.readthedocs.io/en/latest/best-practices/oep-0051-bp-conventional-commits.html)
 - rename `make combine_translations` to `make extract_translations`

**Review by brian-smith-tcril — APPROVED** [review]

File: README.md
Comment (OmarIthawi) [inline README.md:29]:
```suggestion
After this command is run the application can load the translations by changing the device (or the emulator) language in the settings.
```

File: README.md
Comment (OmarIthawi) [inline README.md:34]: Reword and add links:

```suggestion
#### Testing translations
Until the [pull request #422](https://github.com/openedx/openedx-app-ios/pull/422) is merged, translations needs to be pulled from the testing branch `Zeit-Labs/openedx-translations` repo under `fc_55_sample` branch with the following options:

    make ATLAS_OPTIONS='--repository=Zeit-Labs/openedx-translations --revision=fc_55_sample' pull_translations

```

File: README.md
Comment (OmarIthawi) [inline README.md:30]: This section is missing:

```suggestion

#### Using custom translations
 runs , which allows to use custom translations with the following options:

By default the command `make pull_translations` runs [`atlas pull`](https://github.com/openedx/openedx-atlas) which pulls transaltions from the [openedx-translations repository](https://github.com/openedx/openedx-translations).

You can use custom translations on your fork of the openedx-translations repository by setting the following configuration parameters:

- `--revision` (default: `"main"`): Branch or git tag to pull translations from.
- `--repository` (default: `"openedx/openedx-translations"`). GitHub repository slug. There's a feature request to [support GitLab and other providers](https://github.com/openedx/openedx-atlas/issues/20).

Arguments can be passed via the `ATLAS_OPTIONS` environment variable as shown below:

    make ATLAS_OPTIONS='--repository=<your-github-org>/<repository-name> --revision=<branch-name>' pull_translations

There are additional arguments can be passed to `atlas pull`. Refer to the [atlas documentations ](https://github.com/openedx/openedx-atlas) for more information.

```

File: README.md
Comment (OmarIthawi) [inline README.md:40]: The suggested workflow is not correct, please find the edits below: 

```suggestion
### How to translate the app

Translations are managed in the [open-edx/openedx-translations](https://app.transifex.com/open-edx/openedx-translations/dashboard/) Transifex project.

To translate the app join Transifex and add your translations `openedx-app-ios` resource:

 - https://app.transifex.com/open-edx/openedx-translations/openedx-app-ios/ (the link will start working after the [pull request #442](https://github.com/openedx/openedx-app-ios/pull/422) is merged)

Once the resource is both 100% translated and reviewed Transifex will push it to the [openedx-translations](https://github.com/openedx/openedx-translations) repository and developers can use the translations in their app.
```

**Review by OmarIthawi — CHANGES_REQUESTED** [review]
Thanks @Amr-Nash. I've reviewed the README and suggested many edits. Please adopt and revise and let me know when it's ready.

You'll be getting another review regarding the Python code from either me or @iamjazzar today.

**Comment by Amr-Nash** [comment]
@OmarIthawi Thank you for your feedback. It is much appreciated and I am waiting for the script review.

**Review by iamjazzar — CHANGES_REQUESTED** [review]
I've done an initial review, but I haven't tested the functionality yet. The code appears to be clean, but there are numerous nested for loops and some ambiguous names.

File: .gitignore
Comment (Amr-Nash) [reply .gitignore:123]: Done

File: i18n_scripts/translation.py
Comment (Amr-Nash) [reply i18n_scripts/translation.py:1]: Done

**Review by OmarIthawi — COMMENTED** [review]
Added some notes.

File: i18n_scripts/translation.py
Comment (iamjazzar) [reply i18n_scripts/translation.py:40]: ‎In this file‎ we have a lot of path work and there's some duplication in file path handling. Consolidating this logic into helper functions could improve maintainability and can make the code more platform-independent.

   [ translation_file = os.path.join(modules_dir, module, module, 'en.lproj', 'Localizable.strings')](https://github.com/openedx/openedx-app-ios/blob/320548c8a40a0bbb63a99147300e0d68668b4fc0/i18n_scripts/translation.py#L40)
       [ if os.path.isdir(os.path.join(modules_dir, directory))](https://github.com/openedx/openedx-app-ios/blob/320548c8a40a0bbb63a99147300e0d68668b4fc0/i18n_scripts/translation.py#L56)
   [ combined_translation_dir = os.path.join(modules_dir, 'I18N', 'en.lproj')](https://github.com/openedx/openedx-app-ios/blob/320548c8a40a0bbb63a99147300e0d68668b4fc0/i18n_scripts/translation.py#L107)
   [ with open(os.path.join(combined_translation_dir, 'Localizable.strings'), 'w') as f:](https://github.com/openedx/openedx-app-ios/blob/320548c8a40a0bbb63a99147300e0d68668b4fc0/i18n_scripts/translation.py#L109)
   [ lang_parent_dir = os.path.join(modules_dir, 'I18N')](https://github.com/openedx/openedx-app-ios/blob/320548c8a40a0bbb63a99147300e0d68668b4fc0/i18n_scripts/translation.py#L141)
   [ file_path = os.path.join(modules_dir, 'I18N', lang_dir, 'Localizable.strings')](https://github.com/openedx/openedx-app-ios/blob/320548c8a40a0bbb63a99147300e0d68668b4fc0/i18n_scripts/translation.py#L178)
       [ combined_translation_dir = os.path.join(modules_dir, module, module, lang_dir)](https://github.com/openedx/openedx-app-ios/blob/320548c8a40a0bbb63a99147300e0d68668b4fc0/i18n_scripts/translation.py#L204)
       [ with open(os.path.join(combined_translation_dir, 'Localizable.strings'), 'w') as f:](https://github.com/openedx/openedx-app-ios/blob/320548c8a40a0bbb63a99147300e0d68668b4fc0/i18n_scripts/translation.py#L206)

File: Makefile
Comment (Amr-Nash) [reply Makefile:4]: done

File: Makefile
Comment (Amr-Nash) [reply Makefile:10]: As I tested the venv should be activated for each line. So I will consider this as resolved also.

File: i18n_scripts/translation.py
Comment (Amr-Nash) [reply i18n_scripts/translation.py:265]: Nice move, done

File: i18n_scripts/translation.py
Comment (Amr-Nash) [reply i18n_scripts/translation.py]: Done

File: i18n_scripts/translation.py
Comment (Amr-Nash) [reply i18n_scripts/translation.py:151]: Renamed to extract_translations_from_file which I think describes what it does better

File: Makefile
Comment (Amr-Nash) [reply Makefile:13]: Only English translations are pushed from the developer side

**Review by OmarIthawi — COMMENTED** [review]
Posting yesterday's review:

**Comment by Amr-Nash** [comment]
Hello @brian-smith-tcril , 
Oman and Jazzar have reviewed the PR multiple times by now. Could you please give it a look?

File: README.md
Comment (brian-smith-tcril) [inline README.md:30]:
```suggestion
This command runs [`atlas pull`](https://github.com/openedx/openedx-atlas) to download the latest translations files from the [openedx/openedx-translations](https://github.com/openedx/openedx-translations) repository. These files contain the latest translations for all languages. In the [openedx/openedx-translations](https://github.com/openedx/openedx-translations) repository each language's translations are saved as a single file e.g. `I18N/I18N/uk.lproj/Localization.strings` ([example](https://github.com/openedx/openedx-translations/blob/6448167e9695a921f003ff6bd8f40f006a2d6743/translations/openedx-app-ios/I18N/I18N/uk.lproj/Localizable.strings)). After these are pulled, each language's translation file is split into the App's modules e.g. `Discovery/Discovery/uk.lproj/Localization.strings`.
```

File: README.md
Comment (brian-smith-tcril) [inline README.md:36]:
```suggestion
By default, the command `make pull_translations` runs [`atlas pull`](https://github.com/openedx/openedx-atlas) with no arguments which pulls transaltions from the [openedx-translations repository](https://github.com/openedx/openedx-translations).
```

**Review by brian-smith-tcril — COMMENTED** [review]
Overall this is looking great! The code seems reasonable, so assuming it has been tested that part has a :+1: from me!

I left a couple suggestions in the README that I hope will make the instructions a little more clear.

Please let me know if you have any questions!

**Comment by OmarIthawi** [comment]
Thanks @brian-smith-tcril! Yes, the manual testing part is what we focused on.

@Amr-Nash please address Brian notes and let's squash the code once more.

**Comment by Amr-Nash** [comment]
Thanks @brian-smith-tcril for your feedback, 
@OmarIthawi I squashed and pushed again. I think it is ready now

**Review by OmarIthawi — APPROVED** [review]
Thanks @Amr-Nash! I've tested it and it worked locally.

We need an iOS engineer to pull and compile, but overall I think it's ready for merge.

**Review by iamjazzar — APPROVED** [review]

**Comment by OmarIthawi** [comment]
@volodymyr-chekyrta would you mind chekcing this pull request? We'd love a test compile from one of your team so we can continue with the other GitHub Actions pipeline steps.

File: README.md
Comment (OmarIthawi) [inline README.md:28]:
```suggestion
# Note: This will install Python requirements, therefore it is recommended to enable a Python virtual environment
make translation_requirements
make pull_translations
```

**Comment by RawanMatar89** [comment]
after testing the instructions and build the app using Ukraine language on simulator I got 
| SignIn View | Registration view |
|--------|-------|
| Simulator Screenshot - iPhone 15 Pro Max - 2024-06-06 at 12 42 11 | simulator_screenshot_C14D87CE-2F0C-47B4-9FC7-483BCAC84F4F (1)|

**Review by volodymyr-chekyrta — DISMISSED** [review]
Tested ✅
Works fine for me 👍

File: README.md
Comment (volodymyr-chekyrta) [inline README.md:54]: How about removing this section from the README?
It seems that the PR description is sufficient for testing

**Comment by OmarIthawi** [comment]
@Amr-Nash please replace the underscore for iOS.

**Comment by Amr-Nash** [comment]
@OmarIthawi Done, please check.

**Review by brian-smith-tcril — APPROVED** [review]

**Comment by openedx-webhooks** [comment]
@Amr-Nash 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-app-ios&entry.867055334=https://github.com/openedx/openedx-app-ios/pull/422&entry.1484655318=https://github.com/Amr-Nash&entry.752974735=2024-05-03+15:58&entry.1917517419=2024-06-12+19:23&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
